File: backend/app/watcher.py
from __future__ import annotations

import hashlib
import threading
import time
from pathlib import Path

from .config import Settings
from .db import Database
from .parser import parse_pair


def stem_of(path: Path) -> str:
    name = path.name
    for suffix in ("_REPORT.pdf", "_TRACES.xlsm"):
        if name.upper().endswith(suffix.upper()):
            return name[: -len(suffix)]
    return path.stem


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    return digest.hexdigest()


class FolderWatcher:
    def __init__(self, settings: Settings, db: Database):
        self.settings = settings
        self.db = db
        self.stop_event = threading.Event()
        self.thread: threading.Thread | None = None
        # Serializes scans so the background loop and a manual /rescan request
        # cannot run concurrently (double-hashing, duplicate parse subprocesses).
        self.scan_lock = threading.Lock()

    def _file_hash(self, path: Path) -> str:
        """SHA-256 of a source file, reusing the stored digest when size and
        mtime are unchanged — avoids re-reading every file on every scan."""
        stat = path.stat()
        meta = self.db.source_meta(str(path))
        if meta and meta["size_bytes"] == stat.st_size and meta["modified_ns"] == stat.st_mtime_ns:
            return meta["sha256"]
        return sha256(path)

    def start(self) -> None:
        if self.thread and self.thread.is_alive():
            return
        self.thread = threading.Thread(target=self._loop, name="fev-folder-watcher", daemon=True)
        self.thread.start()

    def stop(self) -> None:
        self.stop_event.set()
        if self.thread:
            self.thread.join(timeout=5)

    def _loop(self) -> None:
        while not self.stop_event.is_set():
            try:
                self.scan_once()
            except Exception as exc:
                print(f"watcher scan failed: {exc}", flush=True)
            self.stop_event.wait(self.settings.scan_interval_seconds)

    def scan_once(self) -> None:
        # Lock so the background loop and a manual rescan never overlap.
        with self.scan_lock:
            self._scan_once()

    def _scan_once(self) -> None:
        root = self.settings.watch_folder
        root.mkdir(parents=True, exist_ok=True)
        groups: dict[str, dict[str, Path]] = {}
        for path in root.rglob("*"):
            if not path.is_file():
                continue
            upper = path.name.upper()
            kind = "pdf" if upper.endswith("_REPORT.PDF") else "xlsm" if upper.endswith("_TRACES.XLSM") else None
            if not kind:
                continue
            groups.setdefault(stem_of(path), {})[kind] = path.resolve()

        # Fetch jobs once, not once per stem (avoids O(N^2) over the whole folder).
        jobs_by_stem = {job["stem"]: job for job in self.db.list_jobs()}

        for stem, pair in groups.items():
            pdf = pair.get("pdf")
            xlsm = pair.get("xlsm")
            try:
                pdf_hash = self._file_hash(pdf) if pdf else None
                xlsm_hash = self._file_hash(xlsm) if xlsm else None
                if pdf:
                    self.db.register_source(stem, "pdf", pdf, pdf_hash or "")
                if xlsm:
                    self.db.register_source(stem, "xlsm", xlsm, xlsm_hash or "")
            except OSError as exc:
                # A file vanished mid-scan (e.g. OneDrive sync churn) — skip this
                # pair this cycle instead of aborting the entire scan.
                print(f"watcher: source for {stem} unavailable, skipping: {exc}", flush=True)
                continue
            if not pdf or not xlsm:
                self.db.update_job(
                    stem, "pending_pair", pdf_path=str(pdf) if pdf else None, xlsm_path=str(xlsm) if xlsm else None,
                    pdf_hash=pdf_hash, xlsm_hash=xlsm_hash,
                    message=f"Waiting for {'REPORT.pdf' if not pdf else 'TRACES.xlsm'}",
                )
                continue
            combined_hash = hashlib.sha256(f"{pdf_hash}:{xlsm_hash}".encode()).hexdigest()
            current = jobs_by_stem.get(stem)
            source_unchanged = bool(
                current and current.get("pdf_hash") == pdf_hash and current.get("xlsm_hash") == xlsm_hash
            )
            # A test the engineer deleted stays deleted until its source changes;
            # the watch folder must not silently resurrect it.
            if current and current.get("status") == "deleted" and source_unchanged:
                continue
            # If the source is unchanged and we already hold a fully-parsed record
            # (accepted or quarantined), do not re-run the parser. The parse is
            # deterministic, so re-running wastes a subprocess per scan and — worse
            # — would overwrite any manual edits made to a quarantined record.
            if source_unchanged and current and current.get("status") in ("accepted", "quarantined"):
                existing_test = self.db.get_test(current.get("test_id")) if current.get("test_id") else None
                if existing_test and existing_test.get("units", {}).get("trace"):
                    continue
            self.db.update_job(
                stem, "processing", pdf_path=str(pdf), xlsm_path=str(xlsm),
                pdf_hash=pdf_hash, xlsm_hash=xlsm_hash, message="Parsing FEV report and traces",
            )
            try:
                test = parse_pair(pdf, xlsm)
                existing = self.db.find_identity(test, stem)
                low = list(test.get("lowConfidence", []))
                if existing:
                    field_map = {
                        "project": "project", "cycle": "cycle", "config": "config",
                        "transmission": "transmission", "vnNo": "vnNo",
                    }
                    for flag, field in field_map.items():
                        if flag in low and existing.get(field) not in {None, "", "Unknown"}:
                            test[field] = existing[field]
                            low.remove(flag)
                    for field in ("catalystState", "stt", "startSoc"):
                        if not test.get(field) and existing.get(field) is not None:
                            test[field] = existing[field]
                    test["lowConfidence"] = low
                status = "quarantined" if low else "accepted"
                test_id, replaced = self.db.save_test(
                    test, stem, combined_hash, status,
                    f"parsed with {len(low)} low-confidence field(s)",
                )
                self.db.register_source(stem, "pdf", pdf, pdf_hash or "", test_id)
                self.db.register_source(stem, "xlsm", xlsm, xlsm_hash or "", test_id)
                self.db.update_job(
                    stem, status, pdf_path=str(pdf), xlsm_path=str(xlsm), pdf_hash=pdf_hash,
                    xlsm_hash=xlsm_hash, test_id=test_id,
                    message="Corrected source replaced active record" if replaced else ("Ready for review" if low else "Accepted"),
                )
            except Exception as exc:
                self.db.update_job(
                    stem, "quarantined", pdf_path=str(pdf), xlsm_path=str(xlsm),
                    pdf_hash=pdf_hash, xlsm_hash=xlsm_hash, message=str(exc)[:1000],
                )

File: dashboard/electron/watcher.ts
import fs from 'node:fs'
import path from 'node:path'
import { createHash } from 'node:crypto'
import type { Database } from './db.ts'
import type { ParsePair } from './parsePair.ts'

export interface WatcherSettings {
  watchFolder: string
  scanIntervalSeconds: number
}

export function stemOf(filePath: string): string {
  const name = path.basename(filePath)
  for (const suffix of ['_REPORT.pdf', '_TRACES.xlsm']) {
    if (name.toUpperCase().endsWith(suffix.toUpperCase())) return name.slice(0, -suffix.length)
  }
  return name.replace(/\.[^.]+$/, '')
}

export function sha256File(filePath: string): string {
  const digest = createHash('sha256')
  const fd = fs.openSync(filePath, 'r')
  try {
    const buffer = Buffer.alloc(1024 * 1024)
    let read = 0
    while ((read = fs.readSync(fd, buffer, 0, buffer.length, null)) > 0) {
      digest.update(buffer.subarray(0, read))
    }
  } finally {
    fs.closeSync(fd)
  }
  return digest.digest('hex')
}

export class FolderWatcher {
  settings: WatcherSettings
  db: Database
  parsePair: ParsePair
  timer: NodeJS.Timeout | null
  scanning: boolean

  constructor(settings: WatcherSettings, db: Database, parsePair: ParsePair) {
    this.settings = settings
    this.db = db
    this.parsePair = parsePair
    this.timer = null
    this.scanning = false
  }

  /** The expensive disk read, factored out so tests can spy on it directly
   * rather than on hashFile (which runs once per file every scan regardless
   * of whether it hits the cache below). */
  computeHash(filePath: string): string {
    return sha256File(filePath)
  }

  /** Reuses the stored digest when size and mtime are unchanged. */
  hashFile(filePath: string): string {
    const stat = fs.statSync(filePath, { bigint: true })
    const meta = this.db.sourceMeta(filePath)
    if (meta && meta.size_bytes === Number(stat.size) && meta.modified_ns === String(stat.mtimeNs)) {
      return meta.sha256
    }
    return this.computeHash(filePath)
  }

  start(): void {
    if (this.timer) return
    const tick = async () => {
      try {
        await this.scanOnce()
      } catch (error) {
        console.error('watcher scan failed:', error)
      }
      this.timer = setTimeout(tick, this.settings.scanIntervalSeconds * 1000)
    }
    this.timer = setTimeout(tick, 0)
  }

  stop(): void {
    if (this.timer) clearTimeout(this.timer)
    this.timer = null
  }

  /** Serialized so the background loop and a manual rescan never overlap. */
  async scanOnce(): Promise<void> {
    if (this.scanning) return
    this.scanning = true
    try {
      await this.scan()
    } finally {
      this.scanning = false
    }
  }

  private async scan(): Promise<void> {
    const root = this.settings.watchFolder
    fs.mkdirSync(root, { recursive: true })

    const groups = new Map<string, { pdf?: string; xlsm?: string }>()
    for (const entry of fs.readdirSync(root, { recursive: true, encoding: 'utf-8' })) {
      const full = path.resolve(root, entry)
      if (!fs.statSync(full).isFile()) continue
      const upper = path.basename(full).toUpperCase()
      const kind = upper.endsWith('_REPORT.PDF') ? 'pdf' : upper.endsWith('_TRACES.XLSM') ? 'xlsm' : null
      if (!kind) continue
      const stem = stemOf(full)
      const group = groups.get(stem) ?? {}
      group[kind] = full
      groups.set(stem, group)
    }

    const jobsByStem = new Map(this.db.listJobs().map((job) => [job.stem, job]))

    for (const [stem, pair] of groups) {
      const { pdf, xlsm } = pair
      let pdfHash: string | null = null
      let xlsmHash: string | null = null
      try {
        pdfHash = pdf ? this.hashFile(pdf) : null
        xlsmHash = xlsm ? this.hashFile(xlsm) : null
        if (pdf) this.db.registerSource(stem, 'pdf', pdf, pdfHash ?? '')
        if (xlsm) this.db.registerSource(stem, 'xlsm', xlsm, xlsmHash ?? '')
      } catch (error) {
        // A file vanished mid-scan (OneDrive sync churn) — skip this pair.
        console.error(`watcher: source for ${stem} unavailable, skipping:`, error)
        continue
      }

      if (!pdf || !xlsm) {
        this.db.updateJob(stem, 'pending_pair', {
          pdf_path: pdf ?? null, xlsm_path: xlsm ?? null, pdf_hash: pdfHash, xlsm_hash: xlsmHash,
          message: `Waiting for ${!pdf ? 'REPORT.pdf' : 'TRACES.xlsm'}`,
        })
        continue
      }

      const combinedHash = createHash('sha256').update(`${pdfHash}:${xlsmHash}`).digest('hex')
      const current = jobsByStem.get(stem)
      const sourceUnchanged = Boolean(
        current && current.pdf_hash === pdfHash && current.xlsm_hash === xlsmHash,
      )

      // A test the engineer deleted stays deleted until its source changes.
      if (current && current.status === 'deleted' && sourceUnchanged) continue

      // Unchanged source with a fully-parsed record: do not re-run the parser.
      // It is deterministic, and re-running would overwrite manual edits.
      if (sourceUnchanged && current && ['accepted', 'quarantined'].includes(current.status)) {
        const existingTest = current.test_id ? this.db.getTest(current.test_id) : null
        if (existingTest && existingTest.units?.trace) continue
      }

      this.db.updateJob(stem, 'processing', {
        pdf_path: pdf, xlsm_path: xlsm, pdf_hash: pdfHash, xlsm_hash: xlsmHash,
        message: 'Parsing FEV report and traces',
      })

      try {
        const test = await this.parsePair(pdf, xlsm)
        const existing = this.db.findIdentity(test, stem)
        let low = [...(test.lowConfidence ?? [])]
        if (existing) {
          const fieldMap: Record<string, string> = {
            project: 'project', cycle: 'cycle', config: 'config',
            transmission: 'transmission', vnNo: 'vnNo',
          }
          for (const [flag, field] of Object.entries(fieldMap)) {
            if (low.includes(flag) && ![null, undefined, '', 'Unknown'].includes(existing[field])) {
              test[field] = existing[field]
              low = low.filter((f) => f !== flag)
            }
          }
          for (const field of ['catalystState', 'stt', 'startSoc']) {
            if (!test[field] && existing[field] !== null && existing[field] !== undefined) {
              test[field] = existing[field]
            }
          }
          test.lowConfidence = low
        }
        const status = low.length ? 'quarantined' : 'accepted'
        const { testId: id, replaced } = this.db.saveTest(
          test, stem, combinedHash, status, `parsed with ${low.length} low-confidence field(s)`,
        )
        this.db.registerSource(stem, 'pdf', pdf, pdfHash ?? '', id)
        this.db.registerSource(stem, 'xlsm', xlsm, xlsmHash ?? '', id)
        this.db.updateJob(stem, status, {
          pdf_path: pdf, xlsm_path: xlsm, pdf_hash: pdfHash, xlsm_hash: xlsmHash, test_id: id,
          message: replaced
            ? 'Corrected source replaced active record'
            : low.length ? 'Ready for review' : 'Accepted',
        })
      } catch (error) {
        this.db.updateJob(stem, 'quarantined', {
          pdf_path: pdf, xlsm_path: xlsm, pdf_hash: pdfHash, xlsm_hash: xlsmHash,
          message: String(error).slice(0, 1000),
        })
      }
    }
  }
}
